### PR TITLE
Orchestrator `ResponseDeserializer` codegen and auth

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
This PR refactors the orchestrator's error handling and revises the code generator to implement the new `ResponseDeserializer` trait behind the `enableNewSmithyRuntime` codegen flag.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by ysaito1001 — APPROVED** [review]

File: rust-runtime/aws-smithy-runtime-api/src/type_erasure.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::any::Any;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+
+#[derive(Debug)]
+pub struct TypedBox<T> {
```
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime-api/src/type_erasure.rs:11]: I would love to see docs for this

File: rust-runtime/aws-smithy-http/src/result.rs [hunk 2]
```diff
@@ -124,14 +124,16 @@ pub mod builders {
     /// Builder for [`ServiceError`](super::ServiceError).
     #[derive(Debug)]
     pub struct ServiceErrorBuilder<E, R> {
-        source: Option<E>,
+        modeled_source: Option<E>,
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/result.rs:127]: Is this change still necessary?

File: aws/rust-runtime/aws-http/src/request_id.rs [hunk 3]
```diff
@@ -58,6 +58,12 @@ impl<B> RequestId for http::Response<B> {
     }
 }
 
+impl RequestId for HeaderMap {
+    fn request_id(&self) -> Option<&str> {
+        extract_request_id(self)
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-http/src/request_id.rs:63]: This one will definitely always be fallible.

File: aws/rust-runtime/aws-inlineable/src/s3_request_id.rs [hunk 4]
```diff
@@ -59,6 +59,12 @@ impl<B> RequestIdExt for http::Response<B> {
     }
 }
 
+impl RequestIdExt for HeaderMap {
+    fn extended_request_id(&self) -> Option<&str> {
+        extract_extended_request_id(self)
+    }
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/rust-runtime/aws-inlineable/src/s3_request_id.rs:65]: Yeah, this is the S3 extended request ID.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/protocols/HttpBoundProtocolGenerator.kt [hunk 5]
```diff
@@ -256,6 +403,43 @@ class HttpBoundProtocolTraitImplGenerator(
         val outputSymbol = symbolProvider.toSymbol(outputShape)
         val errorSymbol = symbolProvider.symbolForOperationError(operationShape)
         return protocolFunctions.deserializeFn(operationShape, fnNameSuffix = "http_response") { fnName ->
+            Attribute.AllowClippyUnnecessaryWraps.render(this)
+            rustBlockTemplate(
+                "pub fn $fnName(response: &mut #{http}::Response<#{SdkBody}>) -> std::result::Result<#{O}, #{E}>",
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/protocols/HttpBoundProtocolGenerator.kt:408]: They're in a `pub(crate)` module.

File: rust-runtime/aws-smithy-http/src/result.rs
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-http/src/result.rs:1]: Ended up removing the modeled/unmodeled source changes to `ServiceError` since they're no longer needed.
